### PR TITLE
Add Schwarz subdomain logic

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

A few functions that define overlap regions within the extents of an element for the Schwarz subdomain solver.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                     const DataVector& collocation_points) noexcept {
+  ASSERT(overlap_extent < collocation_points.size(),
+         "Overlap extent is "
+             << overlap_extent
+             << " but must be strictly smaller than the number of grid points ("
+             << collocation_points.size() << ") to compute an overlap width.");
+  for (size_t i = 0; i < collocation_points.size() / 2; ++i) {
+    ASSERT(collocation_points[i] +
+                   collocation_points[collocation_points.size() - i - 1] <
+               1.e-14,
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp]: Use `equal_within_roundoff`.  (The current version will pass if the error is negative.)

File: tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_OverlapHelpers.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    CHECK(overlap_extent(0, 0) == 0);
+    CHECK(overlap_extents(Index<1>{{{3}}}, 2, 0) == Index<1>{{{2}}});
+    CHECK(overlap_extents(Index<2>{{{3, 4}}}, 2, 0) == Index<2>{{{2, 4}}});
+    CHECK(overlap_extents(Index<2>{{{3, 4}}}, 2, 1) == Index<2>{{{3, 2}}});
+    CHECK(overlap_extents(Index<3>{{{3, 4, 5}}}, 2, 0) ==
+          Index<3>{{{2, 4, 5}}});
+    CHECK(overlap_extents(Index<3>{{{3, 4, 5}}}, 2, 1) ==
+          Index<3>{{{3, 2, 5}}});
+    CHECK(overlap_extents(Index<3>{{{3, 4, 5}}}, 2, 2) ==
+          Index<3>{{{3, 4, 2}}});
```
Comment (wthrowe) [inline tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_OverlapHelpers.cpp]: Test a case where the volume extent is the most restrictive limit.

**Review by wthrowe — CHANGES_REQUESTED** [review]
I don't see the documentation rendering, but that seems to be a problem on my end sometimes.  Check that it shows up for you.

File: tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_OverlapHelpers.cpp (hunk 2 above)
Comment (nilsvu) [reply tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_OverlapHelpers.cpp]: I actually realized the `overlap_extents` function is not necessary, so I removed the function in a fixup commit. All it did was replace the volume extent in one of the dimensions with the overlap extent; the interface I'm building will be cleaner without this function. I replaced it with a function `overlap_num_points` that just computes the number of grid points in the overlap region.

**Comment by nilsvu** [comment]
@wthrowe Thanks for reviewing, fixups posted!

**Comment by wthrowe** [comment]
Looks good.  Squash.

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+namespace LinearSolver::Schwarz {
+
+size_t overlap_extent(const size_t volume_extent,
+                      const size_t max_overlap) noexcept {
+  return std::min(max_overlap, volume_extent - 1);
+}
+
+namespace {
+template <size_t Dim>
+void assert_overlap_extent(const Index<Dim>& volume_extents,
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp:24]: why factor these `ASSERT`s when they're used only in one place?

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/// Data structure that can store the `ValueType` on each possible overlap of an
+/// element-centered subdomain with its neighbors. Overlaps are identified by
+/// their `OverlapId`.
+template <size_t Dim, typename ValueType>
+using OverlapMap =
+    FixedHashMap<maximum_number_of_neighbors(Dim), OverlapId<Dim>, ValueType,
+                 boost::hash<OverlapId<Dim>>>;
+
+/*!
+ * \brief The number of points that an overlap extends into the `volume_extent`
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp:31]: What's not clear to me is whether this is the number of points in the volume sense or the number of points in the normal direction to the share face. Could you clarify that in the comment?

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ *   schemes on Gauss-Lobatto grids are limited to the grid points on the
+ *   element face, e.g. for a DG operator that is pre-multiplied by the mass
+ *   matrix, or one where boundary contributions are lifted using the diagonal
+ *   mass-matrix approximation. Not including the grid points facing away from
+ *   the subdomain in the overlap allows to ignore that face altogether in the
+ *   subdomain operator.
+ */
+size_t overlap_extent(size_t volume_extent, size_t max_overlap) noexcept;
+
+/// Number of grid points in an overlap region that extends `overlap_extent`
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp]: I don't understand the difference between `overlap_extent` and this function. Probably related to my question above

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+/// points into the `volume_extents` from either side of the `overlap_dimension`
+template <size_t Dim>
+size_t overlap_num_points(const Index<Dim>& volume_extents,
+                          size_t overlap_extent,
+                          size_t overlap_dimension) noexcept;
+
+/*!
+ * \brief Width of an overlap extending `overlap_extent` points into the
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp:80]: What do you mean by `width`? There is no map passed here, and in 2d/3d the width would be a `DataVector`

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * `collocation_points` from either side.
+ *
+ * Assumes the `collocation_points` are symmetric.
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp]: I have no idea what it means for the collocation points to be symmetric

File: src/ParallelAlgorithms/LinearSolver/Schwarz/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(LIBRARY ParallelSchwarz)
+
+set(LIBRARY_SOURCES
+  OverlapHelpers.cpp
+  )
+
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/CMakeLists.txt:13]: please use the new way of add source files and header files

File: src/ParallelAlgorithms/LinearSolver/Schwarz/Schwarz.hpp [hunk 9]
```diff
@@ -0,0 +1,7 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/Schwarz.hpp:4]: Please add this file to the `CMakeLists.txt`

**Review by nilsdeppe — COMMENTED** [review]
Please go ahead and squash directly

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp (hunk 3 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp:24]: I'm going to use them in #2361 once I rebase that PR on top of this one

**Comment by nilsvu** [comment]
@nilsdeppe squashed in your suggested documentation changes. Thanks for reviewing @wthrowe and @nilsdeppe!

File: src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp (hunk 3 above)
Comment (nilsdeppe) [reply src/ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.cpp:24]: Sounds good, thanks!

**Review by nilsdeppe — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe please look at

**Review by wthrowe — APPROVED** [review]

**Pull request merged**